referenceqvm/stabilizer_utils.py

```diff
@@ -0,0 +1,226 @@
+
+    The classical state is enumerated as the left most bit is the least-significant
+    bit.  This is how one usually reads off classical data from the QVM.  Not
+    how the QVM stores computational basis states.
+
+    :param classical_state: binary repr of a state or an integer.  Should be
+                            left most bit (0th position) is the most significant bit
+    :param num_qubits:
+    :return: new classical state and the coefficient it picked up.
+    """
+    if not isinstance(pauli_operator, PauliTerm):
+        raise TypeError("pauli_operator must be a PauliTerm")
+
+    if not isinstance(classical_state, (list, int)):
+        raise TypeError("classical state must be a list or an integer")
+
+    if isinstance(classical_state, int):
+        if classical_state < 0:
+            raise TypeError("classical_state must be a positive integer")
+
+        classical_state = list(map(int, np.binary_repr(classical_state,
+                                                       width=num_qubits)))
+    if len(classical_state) != num_qubits:
+        raise TypeError("classical state not long enough")
+
+    # iterate through tensor elements of pauli_operator
+    new_classical_state = classical_state.copy()
+    coefficient = 1
+    for qidx, telem in pauli_operator:
+        if telem == 'X':
+            new_classical_state[qidx] = new_classical_state[qidx] ^ 1
+        elif telem == 'Y':
+            new_classical_state[qidx] = new_classical_state[qidx] ^ 1
+            # set coeff
+            if new_classical_state[qidx] == 0:
+                coefficient *= -1j
+            else:
+                coefficient *= 1j
+
+        elif telem == 'Z':
+            # set coeff
+            if new_classical_state[qidx] == 1:
+                coefficient *= -1
+
+    return new_classical_state, coefficient
+
+
+def state_family_generator(state, pauli_operator):
+    """
+    Generate a new state by applying the pauli_operator to each computational bit-string
+
+    This is accomplished in a sparse format where a sparse vector is returned
+    after the action is accumulate in a new list of data and indices
+
+    :param csc_matrix state: wavefunction represented as a column sparse matrix
+    :param PauliTerm pauli_operator: action to apply to the state
+    :return: new state
+    :rtype: csc_matrix
+    """
+    if not isinstance(state, csc_matrix):
+        raise TypeError("we only take csc_matrix")
+
+    num_qubits = int(np.log2(state.shape[0]))
+    new_coeffs = []
+    new_indices = []
+
+    # iterate through non-zero
+    rindices, cindices = state.nonzero()
+    for ridx, cidx in zip(rindices, cindices):
+        # this is so gross looking
+        bitstring = list(map(int, np.binary_repr(ridx, width=num_qubits)))[::-1]
+        new_ket, new_coefficient = compute_action(bitstring, pauli_operator, num_qubits)
+        new_indices.append(int("".join([str(x) for x in new_ket[::-1]]), 2))
```

Comment:
this seems like an odd way of going from bits to an integer. maybe make a function to make it clear what you're doing here. You can avoid converting to string if you do something like

```python
places = 2**np.arange(n)
num = np.sum(places * bits)
```